### PR TITLE
Create Posts “2022-01-29-things-i-learned-in-2021”

### PR DESCRIPTION
**Pull request by nvh95**
Automatically generated by Netlify CMS

### CODE REVIEW

File: content/posts/2022-01-29-things-i-learned-in-2021.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  - rust
+  - wireshark
+  - nextjs
+  - ctf
+  - web security
+  - aws
+---
+
+At the beginning of 2021, I knew that I’m in the middle of Valley of Despair of the [Dunning-Kruger Effect](https://en.wikipedia.org/wiki/Dunning%E2%80%93Kruger_effect). I also realized that I was having a lot of gaps in my knowledge. So I forced myself to learn and practice harder than previous years to fill these holes. In this post, I want to share something that I did and learned through the year, with some context on each item. I think it’s good for me to look back, and it also helps me to plan better this year (2022). And the next reason is that I love sharing 😬
+
```
Comment (minhmo1620) [inline content/posts/2022-01-29-things-i-learned-in-2021.md:22]: You should provide the table of contents here. For example,

To give you a better overview, these are the topics that I would share in the upcoming parts!
- Update my open source
- ....

File: content/posts/2022-01-29-things-i-learned-in-2021.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
... I decided to spend some weekends working on this little project. People started to thank me for helping them. I felt wonderful that my work can save some people’s time. I also learned a lot (I will write a post on what I learned when doing open source). I realized that I actually received more than I gave to the community. In October, I wanted to [release the package to version 2.x](https://github.com/nvh95/react-linkedin-login-oauth2/pull/50) to tackle some major problems in version 1.x. I learned a lot. I converted my code to Typescript, I learned more about Typescript. I used Rollup to bundle my code instead of webpack to support [ECMAScript modules](https://nodejs.org/api/esm.html), to leverage the [Tree Shaking](https://developer.mozilla.org/en-US/docs/Glossary/Tree_shaking) feature. For the first time, I knew about Vite, and use it for the example while developing the package. I had so much fun working on these items and learn so much. After investing some time in the project, I see the strike in the number of weekly downloads that amazed me. It reached about 4000 downloads per week. There are many things to do for this little project and I hope I can make it better in 2022.
```
Comment (minhmo1620) [inline content/posts/2022-01-29-things-i-learned-in-2021.md:29]: You should break this paragraph into 2 paragraphs at `In October, ...`

File: content/posts/2022-01-29-things-i-learned-in-2021.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+## Participated more actively in the open source world
+
+![opensource](/media/2021/opensource.png)
+
+I love open source. I use open source software every-single-day. I use React, Node.js, VS Code and many many other things. Humanity can go that far I think open source plays a big role. The good thing about open source is that everyone can make their lives and others’ lives much easier and more enjoyable. Anyone can contribute to the open source. It does not require you have to write code to core services of a framework or application. It can be anything, from a typo to the recommendations, the feedback about the software. Next year, I target to contribute more to some projects that I like and am interested like msw and Next.js.
```
Comment (minhmo1620) [inline content/posts/2022-01-29-things-i-learned-in-2021.md:35]: `Humanity can go that far I think open source plays a big role.` --> I think open source plays a big role in bringing humanity this far.

Comment (minhmo1620) [reply content/posts/2022-01-29-things-i-learned-in-2021.md:35]: This paragraph is a little bit off when reading the text only. You can change the beginning to `Anyone can contribute to the open source, and so did I`. And then, provide your reasons more after that.

File: content/posts/2022-01-29-things-i-learned-in-2021.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+## Adopt Typescript
+
+![Typescript](/media/2021/typescript.png)
+
+I worked with Javascript every day for more than 4 years. You might heard “Javascript is great”. Yeah, I love it. You might hear “Javascript is terrible”. Well, I... don’t disagree. I knew the good and the bad things about javascript but I still love it overall. It gives me flexibility, but also gives me “Cannot read properties of undefined” 😂. Until one day, Microsoft introduced Typescript. I heard about it and thought: “Ok, javascript with types, just another tech. No thanks, I’m good with pure javascript”. But a few months ago, I decided to experiment Typescript for a greenfield project at my company. I and my teammates were amazed by how Typescript helps us improve the quality of our code and catch almost all the potential bugs early. Most important, Typescript gives us a high level of confidence to ship our code to production. Typescript also gives Github Copilot a lot of context, so Copilot can be really helpful. I have much fun working with both Typescript and Copilot. Typescript is not just a “type system” as I imagined before. It’s truly your best sidekick.
```
Comment (minhmo1620) [inline content/posts/2022-01-29-things-i-learned-in-2021.md:41]: `I and my teammates` --> `My teammates and I`

Comment (minhmo1620) [reply content/posts/2022-01-29-things-i-learned-in-2021.md:41]: `Typescript also gives Github Copilot a lot of context, so Copilot can be really helpful`, hepful for what?

File: content/posts/2022-01-29-things-i-learned-in-2021.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+## Take Web Security course from Standford
+
+![lock in the middle of keys of keyboard](/media/2021/security.jpeg)
+
+Aside from joining a CTF, I also took [a course about Web Security](http://cs253.stanford.edu/). My daily job is developing a web application. Besides making software that helps my users on their jobs, I should also figure out how to protect them from the dangerous threads while they use my software. To achieve that, I need to understand basic and advanced web security concepts. I also need to know how to perform an attack, then I will be able to prevent them. Through the course, [Feross](https://twitter.com/feross) did a great job of guiding learners through these above items. The course helped me write a significantly better code than before.
```
Comment (minhmo1620) [inline content/posts/2022-01-29-things-i-learned-in-2021.md:79]: `To achieve that` --> `To achieve that goal`

File: content/posts/2022-01-29-things-i-learned-in-2021.md (hunk 1 above)
Comment (minhmo1620) [reply content/posts/2022-01-29-things-i-learned-in-2021.md:22]: I also suggest dividing topics based on some specific units such as Learning new things, Open Source exploration, Improve my work experience, etc.
OR
List out based on time!
I think it is also helpful to provide how much time you spend on each task (You can draw a diagram to show a year's improvement).

**Review by minhmo1620 — COMMENTED** [review]
General, I think you did a great job in listing out, I want to hear stories in a better structure so I can't take my eyes off this blog. Great year, good job ❤️

**Comment by nvh95** [comment]
@all-contributors please add @minhmo1620 for review

**Pull request merged**